inlang/source-code/plugins/icu1-importer/src/parse.ts

```diff
@@ -0,0 +1,221 @@
+	let branches: Branch[] = [structuredClone(branch)]
+
+	for (const element of elements) {
+		switch (element.type) {
+			case TYPE.literal: {
+				for (const branch of branches) {
+					branch.pattern.push({ type: "text", value: element.value })
+				}
+				break
+			}
+			case TYPE.argument: {
+				for (const branch of branches) {
+					branch.pattern.push({
+						type: "expression",
+						arg: { type: "variable", name: element.value },
+					})
+				}
+				break
+			}
+			case TYPE.time:
+			case TYPE.number:
+			case TYPE.date: {
+				let options: Option[] = []
+				if (typeof element.style === "string") {
+					options = [{ name: "style", value: { type: "literal", name: element.style } }]
+				}
+				if (typeof element.style === "object" && element.style) {
+					options = Object.entries(element.style.parsedOptions).map(([key, value]) => ({
+						name: key,
+						value: { type: "literal", name: value },
+					}))
+				}
+
+				const fnName = {
+					[TYPE.time]: "time",
+					[TYPE.number]: "number",
+					[TYPE.date]: "date",
+				} as const
+
+				for (const branch of branches) {
+					branch.pattern.push({
+						type: "expression",
+						arg: { type: "variable", name: element.value },
+						annotation: {
+							type: "function",
+							name: fnName[element.type],
+							options,
+						},
+					})
+				}
+				break
+			}
+			case TYPE.plural:
+			case TYPE.select: {
+				const newBranches: Branch[] = []
+
+				for (const [option, optionValue] of Object.entries(element.options)) {
+					let selector: Branch["match"][number] | undefined = undefined
+
+					if (element.type === TYPE.select && option !== "other") {
+						selector = [element.value, undefined, option]
+					}
+
+					if (element.type === TYPE.plural && option !== "other") {
+						if (option.startsWith("=")) {
+							const exact = element.value.slice(1)
+							selector = [element.value, undefined, exact]
+						} else {
+							selector = [element.value, "plural", option]
+						}
+					}
+
+					for (const existingBranch of branches) {
+						const newBranchesForBranch = generateBranches(optionValue.value, {
+							...existingBranch,
+							match: selector ? [...existingBranch.match, selector] : existingBranch.match,
+						})
+						newBranches.push(...newBranchesForBranch)
+					}
+				}
+
+				branches = newBranches
+				break
+			}
+			case TYPE.pound: {
+				for (const branch of branches) {
+					const lastMatch = branch.match.at(-1)
+					if (lastMatch) {
+						branch.pattern.push({
+							type: "expression",
+							arg: {
+								type: "variable",
+								name: lastMatch[0],
+							},
+						})
+					}
+				}
+				break
+			}
+			case TYPE.tag:
+			default: {
+				throw new Error("Unkown element type " + element.type)
+			}
+		}
+	}
+
+	return branches
+}
+
+export function createMessage(
+	messageSource: string,
+	bundleId: string,
+	locale: string
+): MessageNested {
```

Comment:
(minor, can be ignored)

I recommend using the object as argument pattern to: 

- increase readability
- adding properties in the future without bein dependent on the argument order

```ts
createMessage({
  source: string, 
  bundleId: string, 
  locale: string
})

```